src/auth/digest/Config.cc

```diff
@@ -151,85 +156,61 @@ authenticateDigestNonceNew(void)
-    newnonce->noncedata.randomdata = newRandomData(mt);
-
-    authDigestNonceEncode(newnonce);
+    digest_nonce_h::Pointer newnonce = new digest_nonce_h;
+    newnonce->encode(newRandomData(mt));
 
     // ensure temporal uniqueness by checking for existing nonce
-    while (authenticateDigestNonceFindNonce((char const *) (newnonce->key))) {
+    while (NonceCache().find(newnonce->hexKey) != NonceCache().end()) {
         /* create a new nonce */
-        newnonce->noncedata.randomdata = newRandomData(mt);
-        authDigestNonceEncode(newnonce);
+        newnonce->encode(newRandomData(mt));
     }
 
-    hash_join(digest_nonce_cache, newnonce);
-    /* the cache's link */
-    authDigestNonceLink(newnonce);
-    newnonce->flags.incache = true;
+    newnonce->cache();
     debugs(29, 5, "created nonce " << newnonce << " at " << newnonce->noncedata.creationtime);
     return newnonce;
 }
 
 static void
 authenticateDigestNonceSetup(void)
 {
-    if (!digest_nonce_cache) {
-        digest_nonce_cache = hash_create((HASHCMP *) strcmp, 7921, hash_string);
-        assert(digest_nonce_cache);
-        eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
-    }
+    eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
 }
 
 void
 authenticateDigestNonceShutdown(void)
 {
-    /*
-     * We empty the cache of any nonces left in there.
-     */
-    digest_nonce_h *nonce;
-
-    if (digest_nonce_cache) {
-        debugs(29, 2, "Shutting down nonce cache");
-        hash_first(digest_nonce_cache);
-
-        while ((nonce = ((digest_nonce_h *) hash_next(digest_nonce_cache)))) {
-            assert(nonce->flags.incache);
-            authDigestNoncePurge(nonce);
-        }
+    // empty the cache of any nonces left in there
+    for (const auto &entry: NonceCache()) {
+        entry.second->purge();
     }
-
     debugs(29, 2, "Nonce cache shutdown");
 }
 
 static void
 authenticateDigestNonceCacheCleanup(void *)
 {
     /*
-     * We walk the hash by noncehex as that is the unique key we
-     * use.  For big hash tables we could consider stepping through
+     * For big hash tables we could consider stepping through
      * the cache, 100/200 entries at a time. Lets see how it flies
      * first.
      */
-    digest_nonce_h *nonce;
     debugs(29, 3, "Cleaning the nonce cache now");
     debugs(29, 3, "Current time: " << current_time.tv_sec);
-    hash_first(digest_nonce_cache);
 
-    while ((nonce = ((digest_nonce_h *) hash_next(digest_nonce_cache)))) {
-        debugs(29, 3, "nonce entry  : " << nonce << " '" << (char *) nonce->key << "'");
+    for (const auto &entry: NonceCache()) {
+        auto nonce = entry.second;
+        if (!nonce)
```

Comment:
How is a nil cache entry possible? The only code that adds to the cache is the digest_nonce_h::cache() method, right? That method adds "this" so it cannot add a nil pointer. If I am not missing something, we should assert() or assume that the entry is there.